docs/src/apalache/principles/recursive.md

```diff
@@ -1,124 +1,132 @@
-iterations. For instance, consider the following specification:
+A note on (2): In practice, when one fixes specification parameters (that is,
+`CONSTANTS`), it is sometimes possible to find a bound on the number of operator iterations. For instance, consider the following specification:
 
 ```tla
 {{#include ../../../../test/tla/Rec6.tla::}}
 ```
 
-It is clear that the expression `Sum(S)` requires the number of iterations that
-is equal to `Cardinality(S) + 1`. Moreover, the expression `set \subseteq
-1..N` is an invariant, and thus every call `Sum(set)` requires up to `N+1`
-iterations.
+It is clear that the expression `Sum(S)` requires `Cardinality(S)` steps of recursive computation. Moreover, as the unspecified invariant `set \subseteq 1..N` always holds for this specification, every call `Sum(set)` requires up to `N` recursive steps.
 
-When we can find an upper bound on the number of iterations, Apalache can
-unroll the recursive operator up to this bound. To this end, we define two
-additional operators. For instance:
+### The previously supported approach
+
+Previously, when it was possible to find an upper bound on the number of iterations of an operator `Op`, such as `N` for `Sum` in the example above, Apalache would unroll the recursive operator up to this bound. 
+Two additional operators, `UNROLL_DEFAULT_Op` and `UNROLL_TIMES_Op`, were required, for instance:
 
 ```tla
-{{#include ../../../../test/tla/Rec6.tla::}}
+{{#include ../../../../test/tla/Rec6.tla:20:21}}
 ```
 
-In this case, Apalache unrolls every call to `Sum` exactly `UNROLL_TIMES_Sum`
-times, that is, four times. On the default branch, Apalache places
-`UNROLL_DEFAULT_Sum`, that is, 0.
+With the operators `UNROLL_DEFAULT_Op` and `UNROLL_TIMES_Op`, 
+Apalache would internally replace the definition of `Op` with an operator `OpInternal`, that had the following property:
+ 1. `OpInternal` was non-recursive
+ 2. If computing `Op(x)` required a recursion stack of depth at most `UNROLL_TIMES_Op`, then `OpInternal(x) = Op(x)`
+ 3. Otherwise, `OpInternal(x)` would return the value, which would have been produced by the computation of `Op(x)`, if all applications of `Op` while the recursion stack height was `UNROLL_TIMES_Op` returned `UNROLL_DEFAULT_Op` instead of the value produced by another recursive call to `Op`
 
-All recursively defined operators should follow this convention where, for every such operator `Oper`, the user defines both `UNROLL_TIMES_Oper`, which expands to a positive integer value, and `UNROLL_DEFAULT_Oper`, which expands to some default value `Oper(args*)` should take, if the computation would require more than `UNROLL_TIMES_Oper` recursive calls.
-At present, we only support literals (e.g. `4`) or primitive arithmetic expressions (e.g. `2 + 2`) in the body of `UNROLL_TIMES_Oper`.
+Unsurprisingly, (3) caused a lot of confusion, particularly w.r.t. the meaning of the value `UNROLL_DEFAULT_Op`. Consider the following example:
```

Comment:
I would remove this from the  manual if it will no longer be supported. I don't think there's value in recording how things used to work but don't any more?

Reply:
It makes some sense to briefly explain why we don't support recursive operators, imo, but not to give a whole overview of a now unsupported mechanism we abandoned.